e2e: run generic ephemeral for kubernetes 1.21+

Madhu-1 opened the pull request:
Currently, we are skipping the generic ephemeral testing if the kubernetes version is less than 1.21 because of this one the who test suite is getting skipped and e2e is marked as success in 2 minutes. This commit runs the ephemeral tests if the kube=>1.21+. If we do this, for the lower version we can run other tests.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

* sample logs from CI

```log
S [SKIPPING] [39.739 seconds]

cephfs

/go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:185

  Test CephFS CSI

  /go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:273

    Test CephFS CSI [It]

    /go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:274



    generic ephemeral volume only supported from v1.21+



    /go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:319

------------------------------

RBD Upgrade Testing Test RBD CSI 

  Test RBD CSI

  /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:163

[BeforeEach] RBD Upgrade Testing

  /go/src/github.com/ceph/ceph-csi/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:185

STEP: Creating a kubernetes client

Nov 25 12:05:59.289: INFO: >>> kubeConfig: /root/.kube/config

STEP: Building a namespace api object, basename upgrade-test-rbd

STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in upgrade-test-rbd-3274

STEP: Waiting for a default service account to be provisioned in namespace

[BeforeEach] RBD Upgrade Testing

  /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:37

[AfterEach] RBD Upgrade Testing

  /go/src/github.com/ceph/ceph-csi/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:186

Nov 25 12:05:59.439: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready

STEP: Destroying namespace "upgrade-test-rbd-3274" for this suite.

[AfterEach] RBD Upgrade Testing

  /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:102



S [SKIPPING] in Spec Setup (BeforeEach) [0.158 seconds]

RBD Upgrade Testing

/go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:19

  Test RBD CSI [BeforeEach]

  /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:162

    Test RBD CSI

    /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:163



    Skipping RBD Upgrade Testing



    /go/src/github.com/ceph/ceph-csi/e2e/upgrade-rbd.go:39

------------------------------



Ran 0 of 4 Specs in 147.860 seconds

SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 4 Skipped

--- PASS: TestE2E (147.87s)

PASS
```

nixpanic commented on e2e/rbd.go:
```diff
@@ -373,31 +373,30 @@ var _ = Describe("RBD", func() {
-				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
-				if err != nil {
-					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
+				if k8sVersionGreaterEquals(f.ClientSet, 1, 21) {
```
This is not very practical, everything needs to get un-indented once we drop support for Kubernetes < 1.21. This causes more work, more close reviewing and looses git-blame functionality.

Better replace `Skip()` by something else, maybe just log a message and return?

Madhu-1 replied:
return cannot be placed in `By`/`It` block. updated E2E to make use of `Ginkgo.It` instead of `Ginkgo.By`. if we use `ginkgo.It` we can skip the tests the same is not possible in `ginkgo.By`

nixpanic replied:
Ah, that's much better!

Madhu-1 replied:
@nixpanic I reverted back the changes as moving it `It` won't work. currently, this is the easy solution to run E2E for now. but we need to find an alternative way for this one. IMO  this is fine for now instead of skipping the whole test suite.

nixpanic reviewed (commented):
I thought the usage of Skip() like this was not unique, but it seems to be the case...

Madhu-1 commented:
@nixpanic ginkgo `It` cannot be used as a replacement for `By` because `BeforeEach` and `AfterEach` i might need to revert it.

humblec commented:
@Madhu-1 the tests are failing..

Madhu-1 committed:
e2e: run generic ephemeral for kubernetes 1.21+

Currently, we are skipping the generic ephemeral
testing if the kubernetes version is less than
1.21 because of this one the who test suite is
getting skipped and e2e is marked as success
in 2 minutes. This commit runs the ephemeral
tests if the kube=>1.21+. If we do this, for
the lower version we can run other tests.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
> @Madhu-1 the tests are failing..

should be fixed now.

nixpanic reviewed (approved):
this is acceptable for now, but we should consider correcting the e2e structure so that we can use `Skip()` in case we detect some missing features.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Madhu-1 commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

Madhu-1 commented:
>X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

failed due to above error
